File: go/integration_news.go
// integration_news.go
//
// Ports CircleAI.Integration.News:
//   BlueskyOptions / BlueskySource   -> BlueskyOptions / BlueskySource
//   MastodonOptions / MastodonSource -> MastodonOptions / MastodonSource
//   NewsApiOptions / NewsApiSource   -> NewsApiOptions / NewsApiSource
//   RssOptions / RssNewsSource       -> RssOptions / RssNewsSource
//
// Each source is an INewsSource speaking a real HTTP feed API; the live HttpClient
// is replaced by the injected CarrierHTTP seam per the porting rules, so they are
// deterministic and make no network calls. Wire details (URLs, query params,
// auth/UA headers, JSON field extraction, HTML stripping, RSS/Atom parsing, the
// 80-char title truncation with an ellipsis, and the AT-URI → bsky.app URL
// mapping) are reproduced from the C# faithfully.

package circleai

import (
	"context"
	"encoding/xml"
	"errors"
	"strings"
)

// truncateTitle applies the C# "text.Length > 80 ? text[..80] + '…' : text" rule.
// Truncation is by rune (Unicode scalar) — exact for the ASCII/BMP text these
// feeds carry; the appended ellipsis is U+2026.
func truncateTitle(text string) string {
	r := []rune(text)
	if len(r) > 80 {
		return string(r[:80]) + "…"
	}
	return text
}

// ── Bluesky ─────────────────────────────────────────────────────────────────

// blueskyDefaultHost is the C# default AppView host.
const blueskyDefaultHost = "https://public.api.bsky.app"

// BlueskyOptions configures the Bluesky source. Ports BlueskyOptions. An empty
// Host defaults to the public AppView.
type BlueskyOptions struct {
	Query string
	Host  string
}

// BlueskySource reads Bluesky search results over the injected CarrierHTTP. Ports
// BlueskySource.
type BlueskySource struct {
	http CarrierHTTP
	opts BlueskyOptions
}

// NewBlueskySource constructs the source. http is required; an empty Host defaults
// to blueskyDefaultHost.
func NewBlueskySource(http CarrierHTTP, opts BlueskyOptions) (*BlueskySource, error) {
	if http == nil {
		return nil, errors.New("http is required")
	}
	if opts.Host == "" {
		opts.Host = blueskyDefaultHost
	}
	return &BlueskySource{http: http, opts: opts}, nil
}

// SourceID is "bluesky:{query}".
func (s *BlueskySource) SourceID() string { return "bluesky:" + s.opts.Query }

// IsConfigured is true when the Query is non-blank.
func (s *BlueskySource) IsConfigured() bool { return stringsTrimSpaceNonEmpty(s.opts.Query) }

// FetchLatest ports FetchLatestAsync: GET searchPosts?q=&limit=&sort=latest and
// map the "posts" array.
func (s *BlueskySource) FetchLatest(_ context.Context, max int) ([]NewsItem, error) {
	if max <= 0 {
		return nil, errors.New("max out of range")
	}
	url := s.opts.Host + "/xrpc/app.bsky.feed.searchPosts" +
		"?q=" + escapeDataString(s.opts.Query) + "&limit=" + itoaSmall(minInt(max, 100)) + "&sort=latest"
	resp, err := s.http.Do(&CarrierHTTPRequest{Method: "GET", URL: url})
	if err != nil {
		return nil, err
	}
	if !carrierHTTPStatusOK(resp.StatusCode) {
		return nil, statusError("Bluesky searchPosts", resp.StatusCode)
	}
	root, err := parseJSONObject(resp.Body)
	if err != nil {
		return nil, err
	}
	list := []NewsItem{}
	posts, ok := tjArray(root, "posts")
	if !ok {
		return list, nil
	}
	for _, p := range posts {
		post, ok := asJSONObject(p)
		if !ok {
			continue
		}
		uri, _ := tjString(post, "uri")
		record, _ := tjObject(post, "record")
		text := ""
		var ts string
		if record != nil {
			text, _ = tjString(record, "text")
			ts, _ = tjString(record, "createdAt")
		}
		var author string
		if a, ok := tjObject(post, "author"); ok {
			author, _ = tjString(a, "handle")
		}
		tags := []string{}
		if record != nil {
			if facets, ok := tjArray(record, "facets"); ok {
				for _, f := range facets {
					fm, ok := asJSONObject(f)
					if !ok {
						continue
					}
					if feats, ok := tjArray(fm, "features"); ok {
						for _, feat := range feats {
							if featm, ok := asJSONObject(feat); ok {
								if tag, ok := tjString(featm, "tag"); ok {
									tags = append(tags, tag)
								}
							}
						}
					}
				}
			}
		}
		sourceID := author
		if sourceID == "" {
			sourceID = s.SourceID()
		}
		list = append(list, NewsItem{
			ItemID:       uri,
			SourceID:     sourceID,
			Title:        truncateTitle(text),
			Summary:      text,
			URL:          buildBlueskyPostURL(author, uri),
			PublishedUtc: parseDateTimeOffsetUTC(ts),
			Tags:         tags,
		})
	}
	return list, nil
}

// buildBlueskyPostURL maps an AT-URI + handle to a bsky.app post URL. Ports
// BuildPostUrl; a missing handle/uri or a trailing-slash uri yields "about:blank".
func buildBlueskyPostURL(handle, atURI string) string {
	if !stringsTrimSpaceNonEmpty(handle) || !stringsTrimSpaceNonEmpty(atURI) {
		return "about:blank"
	}
	idx := strings.LastIndex(atURI, "/")
	if idx < 0 || idx == len(atURI)-1 {
		return "about:blank"
	}
	rkey := atURI[idx+1:]
	return "https://bsky.app/profile/" + handle + "/post/" + rkey
}

// ── Mastodon ────────────────────────────────────────────────────────────────

// MastodonOptions configures the Mastodon source. Ports MastodonOptions. Hashtag
// and AccessToken are optional (empty == absent).
type MastodonOptions struct {
	Instance    string
	Hashtag     string
	AccessToken string
}

// MastodonSource reads a Mastodon public/hashtag timeline over the injected
// CarrierHTTP. Ports MastodonSource.
type MastodonSource struct {
	http CarrierHTTP
	opts MastodonOptions
}

// NewMastodonSource constructs the source. http is required.
func NewMastodonSource(http CarrierHTTP, opts MastodonOptions) (*MastodonSource, error) {
	if http == nil {
		return nil, errors.New("http is required")
	}
	return &MastodonSource{http: http, opts: opts}, nil
}

// SourceID is "mastodon:{instance}:public" or "mastodon:{instance}:#{hashtag}".
func (s *MastodonSource) SourceID() string {
	if s.opts.Hashtag == "" {
		return "mastodon:" + s.opts.Instance + ":public"
	}
	return "mastodon:" + s.opts.Instance + ":#" + s.opts.Hashtag
}

// IsConfigured is true when the Instance is non-blank.
func (s *MastodonSource) IsConfigured() bool { return stringsTrimSpaceNonEmpty(s.opts.Instance) }

// FetchLatest ports FetchLatestAsync: GET the public or tag timeline and map the
// status array (HTML content stripped to text).
func (s *MastodonSource) FetchLatest(_ context.Context, max int) ([]NewsItem, error) {
	if max <= 0 {
		return nil, errors.New("max out of range")
	}
	var path string
	if s.opts.Hashtag == "" {
		path = "/api/v1/timelines/public?limit=" + itoaSmall(minInt(max, 40))
	} else {
		path = "/api/v1/timelines/tag/" + escapeDataString(s.opts.Hashtag) + "?limit=" + itoaSmall(minInt(max, 40))
	}
	headers := map[string]string{"User-Agent": "CircleAI/1.0 (MastodonSource)"}
	if stringsTrimSpaceNonEmpty(s.opts.AccessToken) {
		headers["Authorization"] = "Bearer " + s.opts.AccessToken
	}
	resp, err := s.http.Do(&CarrierHTTPRequest{Method: "GET", URL: strings.TrimRight(s.opts.Instance, "/") + path, Headers: headers})
	if err != nil {
		return nil, err
	}
	if !carrierHTTPStatusOK(resp.StatusCode) {
		return nil, statusError("Mastodon timeline", resp.StatusCode)
	}
	arr, err := parseJSONArray(resp.Body)
	if err != nil {
		// A non-array body maps to an empty list (C# returns Array.Empty).
		return []NewsItem{}, nil
	}
	list := []NewsItem{}
	for _, raw := range arr {
		st, ok := asJSONObject(raw)
		if !ok {
			continue
		}
		url, _ := tjString(st, "url")
		contentHTML, _ := tjString(st, "content")
		pub, _ := tjString(st, "created_at")
		tags := []string{}
		if tagsArr, ok := tjArray(st, "tags"); ok {
			for _, tg := range tagsArr {
				if tgm, ok := asJSONObject(tg); ok {
					if tn, ok := tjString(tgm, "name"); ok {
						tags = append(tags, tn)
					}
				}
			}
		}
		var acct string
		if a, ok := tjObject(st, "account"); ok {
			acct, _ = tjString(a, "acct")
		}
		text := stripHTMLTags(contentHTML)
		sourceID := acct
		if sourceID == "" {
			sourceID = s.SourceID()
		}
		list = append(list, NewsItem{
			ItemID:       url,
			SourceID:     sourceID,
			Title:        truncateTitle(text),
			Summary:      text,
			URL:          absoluteOrBlank(url),
			PublishedUtc: parseDateTimeOffsetUTC(pub),
			Tags:         tags,
		})
	}
	return list, nil
}

// ── NewsAPI / GNews ─────────────────────────────────────────────────────────

// newsApiDefaultEndpoint is the C# default endpoint.
const newsApiDefaultEndpoint = "https://newsapi.org/v2/everything"

// NewsApiOptions configures the NewsAPI source. Ports NewsApiOptions. An empty
// Endpoint defaults to newsapi.org/v2/everything.
type NewsApiOptions struct {
	APIKey   string
	Query    string
	Endpoint string
}

// NewsApiSource reads a newsapi.org / gnews.io "articles" feed over the injected
// CarrierHTTP. Ports NewsApiSource.
type NewsApiSource struct {
	http CarrierHTTP
	opts NewsApiOptions
}

// NewNewsApiSource constructs the source. http is required; an empty Endpoint
// defaults to newsApiDefaultEndpoint.
func NewNewsApiSource(http CarrierHTTP, opts NewsApiOptions) (*NewsApiSource, error) {
	if http == nil {
		return nil, errors.New("http is required")
	}
	if opts.Endpoint == "" {
		opts.Endpoint = newsApiDefaultEndpoint
	}
	return &NewsApiSource{http: http, opts: opts}, nil
}

// SourceID is "newsapi:{query}".
func (s *NewsApiSource) SourceID() string { return "newsapi:" + s.opts.Query }

// IsConfigured is true when the APIKey is non-blank.
func (s *NewsApiSource) IsConfigured() bool { return stringsTrimSpaceNonEmpty(s.opts.APIKey) }

// FetchLatest ports FetchLatestAsync: GET the endpoint with q/pageSize/sortBy/
// language + X-Api-Key and map the "articles" array. Throws when unconfigured.
func (s *NewsApiSource) FetchLatest(_ context.Context, max int) ([]NewsItem, error) {
	if max <= 0 {
		return nil, errors.New("max out of range")
	}
	if !s.IsConfigured() {
		return nil, errors.New("NewsAPI key not configured.")
	}
	url := s.opts.Endpoint + "?q=" + escapeDataString(s.opts.Query) +
		"&pageSize=" + itoaSmall(minInt(max, 100)) + "&sortBy=publishedAt&language=en"
	headers := map[string]string{
		"X-Api-Key":  s.opts.APIKey,
		"User-Agent": "CircleAI/1.0 (NewsApiSource)",
	}
	resp, err := s.http.Do(&CarrierHTTPRequest{Method: "GET", URL: url, Headers: headers})
	if err != nil {
		return nil, err
	}
	if !carrierHTTPStatusOK(resp.StatusCode) {
		return nil, statusError("NewsAPI everything", resp.StatusCode)
	}
	root, err := parseJSONObject(resp.Body)
	if err != nil {
		return nil, err
	}
	list := []NewsItem{}
	arr, ok := tjArray(root, "articles")
	if !ok {
		return list, nil
	}
	for _, a := range arr {
		art, ok := asJSONObject(a)
		if !ok {
			continue
		}
		title, _ := tjString(art, "title")
		desc, _ := tjString(art, "description")
		url2, _ := tjString(art, "url")
		pub, _ := tjString(art, "publishedAt")
		src := ""
		if sm, ok := tjObject(art, "source"); ok {
			src, _ = tjString(sm, "name")
		}
		sourceID := src
		if sourceID == "" {
			sourceID = s.SourceID()
		}
		list = append(list, NewsItem{
			ItemID:       url2,
			SourceID:     sourceID,
			Title:        title,
			Summary:      desc,
			URL:          absoluteOrBlank(url2),
			PublishedUtc: parseDateTimeOffsetUTC(pub),
			Tags:         []string{},
		})
	}
	return list, nil
}

// ── RSS / Atom ──────────────────────────────────────────────────────────────

// RssOptions configures the RSS/Atom source. Ports RssOptions. SourceID overrides
// the derived host-based id when set.
type RssOptions struct {
	FeedURL  string
	SourceID string
}

// RssNewsSource reads a generic RSS 2.0 / Atom 1.0 feed over the injected
// CarrierHTTP. Ports RssNewsSource.
type RssNewsSource struct {
	http CarrierHTTP
	opts RssOptions
}

// NewRssNewsSource constructs the source. http is required.
func NewRssNewsSource(http CarrierHTTP, opts RssOptions) (*RssNewsSource, error) {
	if http == nil {
		return nil, errors.New("http is required")
	}
	return &RssNewsSource{http: http, opts: opts}, nil
}

// SourceID returns the configured SourceID, or the feed URL's host. Ports the
// getter (`_opts.SourceId ?? _opts.FeedUrl.Host`).
func (s *RssNewsSource) SourceID() string {
	if s.opts.SourceID != "" {
		return s.opts.SourceID
	}
	return uriHost(s.opts.FeedURL)
}

// IsConfigured is always true (ports the C# `=> true`).
func (s *RssNewsSource) IsConfigured() bool { return true }

// FetchLatest ports FetchLatestAsync: GET the feed, parse RSS items then Atom
// entries, concat, and take the first max.
func (s *RssNewsSource) FetchLatest(_ context.Context, max int) ([]NewsItem, error) {
	if max <= 0 {
		return nil, errors.New("max out of range")
	}
	resp, err := s.http.Do(&CarrierHTTPRequest{Method: "GET", URL: s.opts.FeedURL})
	if err != nil {
		return nil, err
	}
	if !carrierHTTPStatusOK(resp.StatusCode) {
		return nil, statusError("RSS feed", resp.StatusCode)
	}
	sourceID := s.SourceID()
	items := append(parseRSS(resp.Body, sourceID), parseAtom(resp.Body, sourceID)...)
	if len(items) > max {
		items = items[:max]
	}
	if items == nil {
		items = []NewsItem{}
	}
	return items, nil
}

// xmlRSS mirrors the RSS 2.0 subset the C# ParseRss reads via
// doc.Descendants("item"). Elements are matched by local name (namespace-agnostic
// for the RSS core, as XName without a namespace does).
type xmlRSS struct {
	XMLName xml.Name    `xml:"rss"`
	Items   []xmlRSSItem `xml:"channel>item"`
}

type xmlRSSItem struct {
	Title       string   `xml:"title"`
	Link        string   `xml:"link"`
	PubDate     string   `xml:"pubDate"`
	Description string   `xml:"description"`
	GUID        string   `xml:"guid"`
	Categories  []string `xml:"category"`
}

// parseRSS ports ParseRss: each <item> → NewsItem (guid ?? link, description
// stripped). A non-RSS document yields no items.
func parseRSS(body []byte, sourceID string) []NewsItem {
	var feed xmlRSS
	if err := xml.Unmarshal(body, &feed); err != nil {
		return nil
	}
	var out []NewsItem
	for _, it := range feed.Items {
		guid := it.GUID
		if guid == "" {
			guid = it.Link
		}
		out = append(out, NewsItem{
			ItemID:       guid,
			SourceID:     sourceID,
			Title:        it.Title,
			Summary:      stripHTMLTags(it.Description),
			URL:          absoluteOrBlank(it.Link),
			PublishedUtc: parseDateTimeOffsetUTC(it.PubDate),
			Tags:         nonNilStrings(it.Categories),
		})
	}
	return out
}

// xmlAtom mirrors the Atom 1.0 subset the C# ParseAtom reads via
// doc.Descendants(atom + "entry"). encoding/xml matches the Atom namespace on the
// element name.
type xmlAtom struct {
	XMLName xml.Name       `xml:"http://www.w3.org/2005/Atom feed"`
	Entries []xmlAtomEntry `xml:"http://www.w3.org/2005/Atom entry"`
}

type xmlAtomEntry struct {
	Title     string          `xml:"http://www.w3.org/2005/Atom title"`
	Links     []xmlAtomLink   `xml:"http://www.w3.org/2005/Atom link"`
	Updated   string          `xml:"http://www.w3.org/2005/Atom updated"`
	Published string          `xml:"http://www.w3.org/2005/Atom published"`
	Summary   string          `xml:"http://www.w3.org/2005/Atom summary"`
	Content   string          `xml:"http://www.w3.org/2005/Atom content"`
	ID        string          `xml:"http://www.w3.org/2005/Atom id"`
	Categories []xmlAtomCategory `xml:"http://www.w3.org/2005/Atom category"`
}

type xmlAtomLink struct {
	Href string `xml:"href,attr"`
}

type xmlAtomCategory struct {
	Term string `xml:"term,attr"`
}

// parseAtom ports ParseAtom: each <entry> → NewsItem (updated ?? published,
// summary ?? content stripped, first link href, category terms).
func parseAtom(body []byte, sourceID string) []NewsItem {
	var feed xmlAtom
	if err := xml.Unmarshal(body, &feed); err != nil {
		return nil
	}
	var out []NewsItem
	for _, e := range feed.Entries {
		link := ""
		if len(e.Links) > 0 {
			link = e.Links[0].Href
		}
		pub := e.Updated
		if pub == "" {
			pub = e.Published
		}
		desc := e.Summary
		if desc == "" {
			desc = e.Content
		}
		guid := e.ID
		if guid == "" {
			guid = link
		}
		tags := []string{}
		for _, c := range e.Categories {
			if c.Term != "" {
				tags = append(tags, c.Term)
			}
		}
		out = append(out, NewsItem{
			ItemID:       guid,
			SourceID:     sourceID,
			Title:        e.Title,
			Summary:      stripHTMLTags(desc),
			URL:          absoluteOrBlank(link),
			PublishedUtc: parseDateTimeOffsetUTC(pub),
			Tags:         tags,
		})
	}
	return out
}

var (
	_ INewsSource = (*BlueskySource)(nil)
	_ INewsSource = (*MastodonSource)(nil)
	_ INewsSource = (*NewsApiSource)(nil)
	_ INewsSource = (*RssNewsSource)(nil)
)

File: c/src/integration_news.c
/*
 * integration_news.c — CircleAI.Integration.News (C11 port).
 *
 * In-memory INewsSource backends for Bluesky / Mastodon / NewsApi / Rss. Each
 * carries the SourceId string the C# property computes from its options, plus an
 * in-memory NewsItem feed seeded via ca_int_news_seed (the injected network
 * data). FetchLatest returns items newest-first, Take(max). Pure C11 + libc.
 * No pthreads.
 */

#include "circle_ai/integration_news.h"
#include "board_common.h"
#include <stdio.h>

typedef struct {
    char               *source_id;   /* owned; the computed SourceId */
    bool                configured;
    bool                gate_config;  /* NewsApi: FetchLatest requires IsConfigured */
    ca_int_news_item_t *items;
    size_t              count, cap;
} news_impl_t;

static const char *news_source_id(void *impl) {
    return ((news_impl_t *)impl)->source_id;
}

static bool news_is_configured(void *impl) {
    return ((news_impl_t *)impl)->configured;
}

/* Stable descending sort of collected indices by PublishedUtc. */
static void news_sort_pub_desc(const news_impl_t *m, size_t *idx, size_t n) {
    for (size_t i = 1; i < n; ++i) {
        size_t key = idx[i];
        int64_t kt = m->items[key].published_utc_ms;
        size_t j = i;
        while (j > 0 && m->items[idx[j - 1]].published_utc_ms < kt) {
            idx[j] = idx[j - 1];
            j--;
        }
        idx[j] = key;
    }
}

static ca_int_news_item_t *news_fetch_latest(void *impl, int max,
                                             size_t *out_count) {
    if (!out_count) return NULL;
    news_impl_t *m = (news_impl_t *)impl;
    if (max <= 0) { *out_count = (size_t)-1; return NULL; }
    if (m->gate_config && !m->configured) { /* NewsApi InvalidOperationException */
        *out_count = (size_t)-1;
        return NULL;
    }
    if (m->count == 0) { *out_count = 0; return NULL; }

    size_t *idx = (size_t *)malloc(m->count * sizeof(size_t));
    if (!idx) { *out_count = (size_t)-1; return NULL; }
    for (size_t i = 0; i < m->count; ++i) idx[i] = i;
    news_sort_pub_desc(m, idx, m->count);
    size_t n = m->count;
    if ((size_t)max < n) n = (size_t)max;

    ca_int_news_item_t *out = (ca_int_news_item_t *)calloc(n, sizeof(*out));
    if (!out) { free(idx); *out_count = (size_t)-1; return NULL; }
    for (size_t i = 0; i < n; ++i) {
        if (!ca_int_news_item_copy(&out[i], &m->items[idx[i]])) {
            ca_int_news_item_free_array(out, i);
            free(idx);
            *out_count = (size_t)-1;
            return NULL;
        }
    }
    free(idx);
    *out_count = n;
    return out;
}

int ca_int_news_seed(ca_int_news_source_t *s, const ca_int_news_item_t *item) {
    if (!s || !item) return -1;
    news_impl_t *m = (news_impl_t *)s->impl;
    ca_int_news_item_t copy;
    if (!ca_int_news_item_copy(&copy, item)) return -1;
    if (m->count == m->cap) {
        size_t nc = m->cap ? m->cap * 2 : 4;
        void *nb = realloc(m->items, nc * sizeof(*m->items));
        if (!nb) { ca_int_news_item_free(&copy); return -1; }
        m->items = (ca_int_news_item_t *)nb;
        m->cap = nc;
    }
    m->items[m->count++] = copy;
    return 0;
}

/* ── construction ───────────────────────────────────────────────────────── */

/* Build a source over an owned source_id string (takes ownership). NULL on OOM
 * (frees source_id). */
static ca_int_news_source_t *news_new(char *source_id, bool configured,
                                      bool gate_config) {
    if (!source_id) return NULL;
    news_impl_t *m = (news_impl_t *)calloc(1, sizeof(news_impl_t));
    if (!m) { free(source_id); return NULL; }
    m->source_id   = source_id;
    m->configured  = configured;
    m->gate_config = gate_config;

    ca_int_news_source_t *s = (ca_int_news_source_t *)calloc(1, sizeof(*s));
    if (!s) { free(source_id); free(m); return NULL; }
    s->impl          = m;
    s->source_id     = news_source_id;
    s->is_configured = news_is_configured;
    s->fetch_latest  = news_fetch_latest;
    return s;
}

/* asprintf-free concat helper: "{prefix}{a}". NULL on OOM. */
static char *cat2(const char *prefix, const char *a) {
    size_t la = a ? strlen(a) : 0, lp = strlen(prefix);
    char *r = (char *)malloc(lp + la + 1);
    if (!r) return NULL;
    memcpy(r, prefix, lp);
    if (la) memcpy(r + lp, a, la);
    r[lp + la] = '\0';
    return r;
}

ca_int_news_source_t *ca_int_bluesky_source_create(const char *query) {
    char *sid = cat2("bluesky:", query ? query : "");
    bool configured = !cab_is_ws(query);
    return news_new(sid, configured, false);
}

ca_int_news_source_t *ca_int_mastodon_source_create(const char *instance,
                                                    const char *hashtag) {
    const char *inst = instance ? instance : "";
    char *sid;
    if (hashtag == NULL || hashtag[0] == '\0') {
        /* "mastodon:{instance}:public" */
        size_t need = strlen("mastodon:") + strlen(inst) + strlen(":public") + 1;
        sid = (char *)malloc(need);
        if (sid) snprintf(sid, need, "mastodon:%s:public", inst);
    } else {
        /* "mastodon:{instance}:#{hashtag}" */
        size_t need = strlen("mastodon:") + strlen(inst) + 2 + strlen(hashtag) + 1;
        sid = (char *)malloc(need);
        if (sid) snprintf(sid, need, "mastodon:%s:#%s", inst, hashtag);
    }
    bool configured = !cab_is_ws(instance);
    return news_new(sid, configured, false);
}

ca_int_news_source_t *ca_int_newsapi_source_create(const char *api_key,
                                                   const char *query) {
    char *sid = cat2("newsapi:", query ? query : "");
    bool configured = !cab_is_ws(api_key);
    return news_new(sid, configured, true); /* FetchLatest gated on IsConfigured */
}

ca_int_news_source_t *ca_int_rss_source_create(const char *feed_host,
                                               const char *source_id) {
    /* SourceId := source_id ?? feed_host (C# _opts.SourceId ?? _opts.FeedUrl.Host). */
    const char *pick = source_id ? source_id : (feed_host ? feed_host : "");
    char *sid = cab_strdup_empty(pick);
    return news_new(sid, true, false); /* Rss IsConfigured is always true */
}

void ca_int_news_source_destroy(ca_int_news_source_t *s) {
    if (!s) return;
    news_impl_t *m = (news_impl_t *)s->impl;
    if (m) {
        free(m->source_id);
        for (size_t i = 0; i < m->count; ++i) ca_int_news_item_free(&m->items[i]);
        free(m->items);
        free(m);
    }
    free(s);
}
